File: src/main/kotlin/at/fhhgb/mc/component/utility/Player.java
package at.fhhgb.mc.component.utility;

import at.fhhgb.mc.component.sim.model.recorder.network.DataSnapshot;
import at.fhhgb.mc.component.sim.model.recorder.network.WLanData;
import at.fhhgb.mc.interfaces.PlayBackEnum;
import at.fhhgb.mc.interfaces.PlaybackCallbackInterface;

import java.util.List;

public class Player {

    final PlayBackEnum playBackEnum;
    List<DataSnapshot> dataSnapshots;
    PlaybackCallbackInterface callback;
    private long playBackSpeed = 500;
    private int[] wayPointCount;

    public Player(List<DataSnapshot> dataSnapshots, PlaybackCallbackInterface callback, PlayBackEnum playBackEnum){
       this.dataSnapshots = dataSnapshots;
        this.callback = callback;
        this.playBackEnum = playBackEnum;
        setNumberOfAPSteps();
    }

    private void setNumberOfAPSteps(){
        wayPointCount = new int[dataSnapshots.get(dataSnapshots.size()-1).getWayPointID()];
        for(DataSnapshot dataSnapshot : dataSnapshots){
            wayPointCount[dataSnapshot.getWayPointID()-1]+=1;
        }
        callback.wayPointCount(wayPointCount);
    }

    public void setPlayBackSpeed(long playBackSpeed) {
        this.playBackSpeed = playBackSpeed;
    }

    public boolean startPlayback(){
        System.out.println(dataSnapshots.size());
        int i = 1;
        int signallevel = -1000;
        String bssid = "";
        for(DataSnapshot dataSnapshot : dataSnapshots){
            System.out.println("Snapshot "+i);
            i++;
            for(WLanData wLanData : dataSnapshot.getwLanDataList()){
                if(wLanData.getSSID().equals("fhhgb")){
                   /* if(playBackEnum == PlayBackEnum.SIMPLE){
                        List<AccessPoint> accessPoints = wLanData.getAccessPoints();
                        Collections.sort(accessPoints);
                        Collections.reverse(accessPoints);
                        callback.nearestAccessPoint(accessPoints);
                    }else if(playBackEnum == PlayBackEnum.WEIGHTED_DISTANCES){
                        callback.allAccessPoints(wLanData.getAccessPoints());
                    }*/
                    //TODO: REMOVE IF IT WORKS

                    callback.allAccessPoints(wLanData.getAccessPoints());
                }
            }
            try {
                Thread.sleep(10);
//                if(playBackSpeed >1){
//                    Thread.sleep(500/playBackSpeed);
//                }else{
//                    Thread.sleep(500*((playBackSpeed-2)*-1));
//                }
            } catch (InterruptedException e) {
            }
        }
        return false;
    }
}

File: src/main/kotlin/at/fhooe/mc/wifipositioning/utility/Player.kt
package at.fhooe.mc.wifipositioning.utility

import at.fhooe.mc.wifipositioning.model.simulation.recording.DataSnapshot
import at.fhooe.mc.wifipositioning.interfaces.PlayBackEnum
import at.fhooe.mc.wifipositioning.interfaces.PlaybackCallbackInterface

class Player(internal var dataSnapshots: List<DataSnapshot>, internal var callback: PlaybackCallbackInterface, internal val playBackEnum: PlayBackEnum) {
    private var playBackSpeed: Long = 1
    private var wayPointCount: IntArray? = null
    var isRunning = false
    private set
    var pausePlayback = false
    var stopPlayback = false

    init {
        setNumberOfAPSteps()
    }

    private fun setNumberOfAPSteps() {
        wayPointCount = IntArray(dataSnapshots[dataSnapshots.size - 1].wayPointID)
        wayPointCount?.let { wayPointCount ->
            for (dataSnapshot in dataSnapshots) {
                wayPointCount[dataSnapshot.wayPointID - 1] += 1
            }
            callback.wayPointCount(wayPointCount)
        }
    }

    fun setPlayBackSpeed(playBackSpeed: Long) {
        this.playBackSpeed = playBackSpeed
    }

    fun startPlayback() {
        println(dataSnapshots.size)
        isRunning = true
        var i = 1
        for (dataSnapshot in dataSnapshots) {

            while (pausePlayback) {
                try {
                    Thread.sleep(1000)
                } catch (e: InterruptedException) {}
            }

            if (stopPlayback) {
                isRunning = false
                pausePlayback = false
                stopPlayback = false
                return
            }

            println("Snapshot $i")
            i++

            dataSnapshot.getwLanDataList()
                    .stream()
                    .filter { wd -> wd.ssid == "fhhgb" }
                    .forEach {
                        callback.allAccessPoints(it.scannedAccessPoints)
                    }

            try {
                Thread.sleep(playBackSpeed)
            } catch (e: InterruptedException) {
            }

        }

        isRunning = false
        pausePlayback = false
        return
    }
}
